### ISSUE TITLE
Add browser-tabs-lock package for analytics app

### ISSUE THREAD

**Issue by erickakoyama** [liferay/liferay-frontend-guidelines#136]
- [X] Before making a proposal, have you used the issue search functionality?

## What is your proposal?
Add [browser-tabs-lock](https://www.npmjs.com/package/browser-tabs-lock) to [analytics js client](https://github.com/liferay/liferay-portal/tree/master/modules/apps/analytics/analytics-client-js) dependencies. It is a package for simulating mutual exclusion when accessing localStorage across browsers.

## Why would adopting this proposal be beneficial?
This dependency is specifically for the [analytics js client](https://github.com/liferay/liferay-portal/tree/master/modules/apps/analytics/analytics-client-js). Using this library will ensure, to a reasonable degree, that no two client processes access and mutate localStorage at the same time.

## What are the alternatives to this proposal?
An alternative could be something like using the [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API) or [Web Locks API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Locks_API); however, the cross-browser support is not complete for either of those options at this time. [Can I use IndexedDB?](https://caniuse.com/#feat=indexeddb).

**Comment by bryceosterhaus** [comment]
[bundlephobia for reference](https://bundlephobia.com/result?p=browser-tabs-lock@1.2.8)

I also cross checked with our [Inbound licensing policy](https://grow.liferay.com/excellence/Liferay+Inbound+Licensing+Policy#mit--mit-license) and since it is MIT, this seems to be fine for DXP.

@wincent any concerns with this?

**Comment by erickakoyama** [comment]
I also wanted to note that although the analytics js client is in the portal repo, it is not used in DXP, but is loaded from a cdn on page load: https://analytics-js-cdn.liferay.com

**Comment by wincent** [comment]
TL;DR:

> I see no problem with adding this as a dependency to `analytics-js-client`

---

Observations about the package:

- [Zero (non-dev) dependencies](https://www.npmjs.com/package/browser-tabs-lock?activeTab=dependencies)  🎉 
- Usage:
  - NPM page is showing 52k downloads in last week
  - But not super broadly used/watched; eg:
    - [Only 10 packages directly depend on this one](https://www.npmjs.com/browse/depended/browser-tabs-lock)
    - [GitHub repo has 7 stars](https://github.com/supertokens/browser-tabs-lock), and has had 5 [issues](https://github.com/supertokens/browser-tabs-lock/issues) and 5 [PRs](https://github.com/supertokens/browser-tabs-lock/pulls) in its lifetime
- Like you said, @bryceosterhaus, it's MIT

There is not much to this package at all; it's only 3 files:

- https://github.com/supertokens/browser-tabs-lock/blob/master/index.ts (267 lines)
- https://github.com/supertokens/browser-tabs-lock/blob/master/bundleEntry.ts (7 lines)
- https://github.com/supertokens/browser-tabs-lock/blob/master/processLock.ts (61 lines)

Something as lightweight as this, I probably would have just written it myself (insert [standard anti-dependency screed here](https://github.com/liferay/liferay-frontend-guidelines/blob/master/dxp/dev_dependencies.md)), but I see no problem with adding this as a dependency to `analytics-js-client`. While we like to vet dependencies, there is a hierarchy here (ranging from lightest vetting to heaviest):

1. **Dependencies in independent projects:** No centralized review flow for these (`analytics-js-client` almost qualifies as an "independent project"; it's a bit of an odd beast because it lives in liferay-portal and it affects the global `yarn.lock`, but it is released separately).
2. **Non-dev dependencies in liferay-portal projects:** No formal review flow here, yet, but we do keep an eye on the `yarn.lock` to try and stop entropy from eventually crushing us — it is likely that we'll start increasing the vetting as liferay-portal becomes more frontend-heavy.
3. **Dev dependencies in liferay-portal:** we vet these so strictly we even have a lint rule stopping them from being added. We started here because this was where the uniformity and control gave us the biggest immediate payoff (in terms of having a standard build/test environment that we could add features to, debug etc).

**Comment by wincent** [comment]
@erickakoyama: Seems like you, @bryceosterhaus, and I are all fine with this going in. Even @jbalsas agrees...

<img width="267" alt="Add_browser-tabs-lock_package_for_analytics_app_·_Issue__136_·_liferay_liferay-frontend-guidelines" src="https://user-images.githubusercontent.com/7074/77531894-023d2c00-6e94-11ea-82f1-31a7636c1460.png">

...so let's close this. Feel free to go ahead and add it to [the project-local dependencies](https://github.com/liferay/liferay-portal/blob/d2f64d83a073182866857d86bc9b06c3555fccc2/modules/apps/analytics/analytics-client-js/package.json).
